### DISCUSSION TITLE
Help with missingness_chart()

### DISCUSSION THREAD

**Discussion by pkky22** [moj-analytical-services/splink#1247] [General]
Hi,

Please bear with my beginner level programming skill and help on the following.

I am trying to replicate splink demo available on github. I am successfully doing many things from the demo except charting.  To start with, missingness_chart() just spits json looking output rather than a chart. My working code is:
```
import pandas as pd 
import altair as alt
alt.renderers.enable('default')

df = pd.read_csv(r~\fake_1000_from_splink_demos.csv")
from splink.duckdb.duckdb_linker import DuckDBLinker
linker = DuckDBLinker(df)
linker.missingness_chart()
```
output: 
```
{'config': {'view': {'continuousWidth': 400,
   'continuousHeight': 300,
   'width': 400},
  'axis': {'labelFontSize': 11}},
 'title': '',
 'layer': [{'mark': 'bar',
   'encoding': {'color': {'type': 'quantitative',
     'field': 'null_proportion',
     'legend': {'format': '.0%'},
     'scale': {'range': 'heatmap'},
     'title': 'Missingness'},
    'tooltip': [{'type': 'nominal', 'field': 'column_name', 'title': 'Column'},
     {'type': 'quantitative',
      'field': 'null_count',
      'format': ',.0f',
      'title': 'Count of nulls'},
     {'type': 'quantitative',
      'field': 'null_proportion',
      'format': '.2%',
      'title': 'Percentage of nulls'},
     {'type': 'quantitative',
      'field': 'total_record_count',
      'format': ',.0f',
      'title': 'Total record count'}],
    'x': {'type': 'quantitative',
     'axis': {'format': '%', 'title': 'Percentage of nulls'},
     'field': 'null_proportion'},
    'y': {'type': 'nominal',
     'axis': {'title': ''},
     'field': 'column_name',
     'sort': '-x'}},
   'title': 'Missingness per column out of 1,000 records'}],
 'data': {'values': [{'null_proportion': 0.0,
    'null_count': 0,
    'total_record_count': 1000,
    'column_name': 'None'},
   {'null_proportion': 0.0,
    'null_count': 0,
    'total_record_count': 1000,
    'column_name': 'unique_id'},
   {'null_proportion': 0.10000002384185791,
    'null_count': 100,
    'total_record_count': 1000,
    'column_name': 'first_name'},
   {'null_proportion': 0.07700002193450928,
    'null_count': 77,
    'total_record_count': 1000,
    'column_name': 'surname'},
   {'null_proportion': 0.0,
    'null_count': 0,
    'total_record_count': 1000,
    'column_name': 'dob'},
   {'null_proportion': 0.10900002717971802,
    'null_count': 109,
    'total_record_count': 1000,
    'column_name': 'city'},
   {'null_proportion': 0.1119999885559082,
    'null_count': 112,
    'total_record_count': 1000,
    'column_name': 'email'},
   {'null_proportion': 0.0,
    'null_count': 0,
    'total_record_count': 1000,
    'column_name': 'group'}],
  'name': 'data-0e7bce5a1d2f132e282789d6ef7780fe'},
 '$schema': 'https://vega.github.io/schema/vega-lite/v4.8.1.json'}
```
Any ideas to fix the issue here?

Thanks!

**Comment by RobinL** [comment]
When using Splink to generate a chart, it will by default render the chart only in interactive environments. These include Jupyter notebooks or the interactive IPython kernel within Visual Studio Code. However, if you're running them as a Python script, the charting functions will only produce a JSON representation of the chart.

If you wish to save the chart as an HTML file that can be viewed offline, assign the chart to a variable. Following this, utilize the `save_offline_chart` function. More information about this function can be found at the following URL:
https://github.com/moj-analytical-services/splink/blob/aba0f5f2ba40958ee642f69df836c1e53cacf9c1/splink/linker.py#L2436

Also note that if Altair is not installed, all charting functions will return the json representation rather than the Altair chart

**Reply by mashby1966** [reply]
Hi,
I have tried to run the same code that pkky22 has shown in both Jupyter Notebook and Visual Studio Code however both only produce a JSON representation.

What should I be checking to get this working in Visual Studio Code

Regards

**Reply by RobinL** [reply]
Does a basic Altair chart display correctly?  e.g https://altair-viz.github.io/gallery/simple_bar_chart.html

**Reply by mashby1966** [reply]
Yes it does

**Reply by RobinL** [reply]
Thanks. What version of Altair do you have installed?

**Reply by mashby1966** [reply]
I have version 5.0.1 installed.

BTW - (not sure if it is relevant) if I use the line "save_offline_chart(c.spec,"test_chart.html",overwrite=True)"` as suggested, I get a "'dict' object has no attribute 'spec'" error however if I use "save_offline_chart(c,"test_chart.html",overwrite=True)" it works and I  can view the test_chart.html in a browser.

Thanks

**Reply by RobinL** [reply]
Thanks I wondered whether that maybe the case.  There have been quite significant changes in the Altair 5 release.  It may be that downgrading Altair to a 4.x version helps?
pip install altair==4.2.2
If that works we will want to update our dependencies to specify Altair < 5 for the moment, although in the longer run there's an open issue for us to upgrade everything to Altair 5

**Reply by mashby1966** [reply]
Perfect, I have just done exactly that before I saw your reply.
The missingness chart is now rendering in VS Code and also the c.spec error is gone.

Thanks

**Reply by RobinL** [reply]
Thanks for confirming!

**Reply by RobinL** [reply]
This is now fixed by #1306 so will be included in the next release. Specifically once released a`pip install splink` will install a version 4 of Altair.  Thanks again for your help 🙏

**Comment by pkky22** [comment]
Thanks for the information. I tried to get Splink generate and render chart in both notebooks and the interactive IPython kernel within Visual Studio Code, but it only produced JSON outputs.

In Jupyter notebook, save_offline_chart works for me as I can refer to the HTML outside of the notebook now.

Thank you!

**Reply by staropram** [reply]
I actually had a similar issue which was related to the vegalite extension. After some fiddling around I actually couldn't get it to work without wrapping the chart output in a `VegaLite(chart)` call and importing `from vega import VegaLite`.

Also, I wrote a wrapper to work from ipython:

```import webbrowser
import uuid

def viewSplinkChart(splinkChart):
    splinkChart['width'] = 1000
    splinkChart['height'] = 500
    out = "tmp/" + str(uuid.uuid4()) + ".html"
    save_offline_chart(splinkChart, out,overwrite=True)
    webbrowser.open(out,new=1)
```

**Comment by pkky22** [comment]
My bad, I should have mentioned before - to save the offline chart, I did use "save_offline_chart(c,"test_chart.html",overwrite=True)"

**Reply by RobinL** [reply]
Yeah, it's the recently released version 5 of Altair that's causing the problem. This is now fixed by #1306 so will be included in the next release. Specifically once released a`pip install splink` will install a version 4 of Altair.

**Comment by mahakti** [comment]
Hi Can anyone help me resolve below error please?

Whenever I am running below code in jupyter-
model = linker.model
model.all_charts_write_html_file(filename=f"../OUT/charts/splink_charts.html", overwrite=True)

I am getting error like - 
UnicodeEncodeError Traceback (most recent call last)
Cell In[25], line 2
      1 model = linker.model
----> 2 model.all_charts_write_html_file(filename=f"../OUT/charts/splink_charts_{day}.html", overwrite=True)

File ~\PycharmProjects\pythonProject1\venv1\lib\site-packages\splink\model.py:257, in Model.all_charts_write_html_file(self, filename, overwrite)
    253 fmt_dict["proportion_of_matches"] = self.current_settings_obj[
    254     "proportion_of_matches"
    255 ]
    256 with open(filename, "w") as f:
--> 257     f.write(template.format(**fmt_dict))

File C:\Users\2202055\AppData\Local\Programs\Python\Python39\lib\encodings\cp1252.py:19, in IncrementalEncoder.encode(self, input, final)
     18 def encode(self, input, final=False):
---> 19     return codecs.charmap_encode(input,self.errors,encoding_table)[0]

UnicodeEncodeError: 'charmap' codec can't encode character '\u2212' in position 23084: character maps to <undefined>


But when I am running code for getting other charts like below, charts are getting displayed without any error in jupyter-
model.probability_distribution_chart()
model.bayes_factor_chart()

Why are individual chart's code running properly but code for making all charts giving error? How can I resolve it?

I am uisng altair version 4, alt.renderers.enable("default")

**Reply by RobinL** [reply]
What version of Splink are you using?

**Reply by mahakti** [reply]
I tried with Splink  version 2.1.14 and also with Splink 2.1.0rc1.

**Reply by mahakti** [reply]
Also when I tried saving individual charts into an html file, I am getting same 'charmap' encoding error. 

Code:
from splink import charts
altair_chart=model.probability_distribution_chart()
charts.save_offline_chart(altair_chart, filename='my_chart.html', overwrite=True, print_msg=True)

Error:
---------------------------------------------------------------------------
UnicodeEncodeError   Traceback (most recent call last)
Cell In[32], line 1
----> 1 charts.save_offline_chart(altair_chart, filename='my_chart.html', overwrite=True, print_msg=True)

File ~\PycharmProjects\pythonProject1\venv1\lib\site-packages\splink\charts.py:81, in save_offline_chart(altair_chart, filename, overwrite, print_msg)
     78 fmt_dict["mychart"] = _make_json(altair_chart)
     80 with open(filename, "w") as f:
---> 81     f.write(template.format(**fmt_dict))
     83 if print_msg:
     84     print(f"Chart saved to {filename}")

File C:\Users\2202055\AppData\Local\Programs\Python\Python39\lib\encodings\cp1252.py:19, in IncrementalEncoder.encode(self, input, final)
     18 def encode(self, input, final=False):
---> 19     return codecs.charmap_encode(input,self.errors,encoding_table)[0]

UnicodeEncodeError: 'charmap' codec can't encode character '\u2212' in position 23084: character maps to <undefined>

However, when I run below code in jupyter, it does give the chart in jupyter. 
model.probability_distribution_chart()

**Reply by RobinL** [reply]
I'm afraid Splink 2.x is now no longer supported, since 3.x is now almost a year old.  I think this issue is now fixed, but only in 3.x, see here https://github.com/moj-analytical-services/splink/pull/562

**Reply by mahakti** [reply]
ok so does this means that there is no way I can save charts using splink2? Any workaround? Please suggest

**Reply by RobinL** [reply]
If you take a look at the changes in [the pr](https://github.com/moj-analytical-services/splink/pull/562/files) and make similar changes to your copy of Splink 2, then it might work.  But generally we recommend upgrading to version 3.  Also worth noting these charts do usually work in Splink 2 (they're not always broken, just occasionally) - it could be something to do with the encoding of your data.

**Reply by mahakti** [reply]
Thanks I will try as you suggested. Also it would be great help if you can tell whether method for linking two datasets is present in version 2 or not? Can I give "link_type": "link_only" in settings, will it work? How to pass both datasets as argument to splink 2?  Is there any documentation of version 2 present?

**Reply by RobinL** [reply]
There's no documentation site for Splink 2 I'm afraid.  The best place to look would be [here](https://github.com/moj-analytical-services/splink_demos/tree/d391cd32128cd0d51e4a1275ea7d833f2c043462), specifically [this notebook](https://github.com/moj-analytical-services/splink_demos/blob/d391cd32128cd0d51e4a1275ea7d833f2c043462/quickstart_demo_linking.ipynb) for linking.

But I would recommend updating your code to Splink 3 if at all possible, it's faster, better supported, and has much more functionality
